[11.0 P4] Blazor What's New feature coverage (#37133)
dotnet/AspNetCore.Docs: aspnetcore/blazor/state-management/server.md

This feature is useful in the following scenarios:

* Planned shutdowns and deployments.
* Instance draining.
* App maintenance windows.

`Circuit.RequestCircuitPauseAsync(CancellationToken)` is used to request that the connected client begin the graceful circuit-pause flow. The `CancellationToken` cancels the request before it is accepted by the framework. The method returns `true` if the request was accepted and the client was asked to begin pausing.

<!-- UPDATE 11.0 - REVIEWER NOTE ... The following example might not be what we show.
                                     It's placed here as a possible example 
                                     based on Javier's original non-RequestCircuitPauseAsync
                                     approach from the issue. If we use this example,
                                     we need some changes to this.
                       
When a server-side Blazor application shuts down (for example, during deployment), connected clients lose their SignalR connection. The approach in this section:

* Detects shutdown before the server closes connections.

> * Triggers a pause on connected circuits.

Reviewer: ```suggestion
* Triggers a pause on connected circuits via <xref:Microsoft.AspNetCore.Components.Server.Circuits.Circuit.RequestCircuitPauseAsync%2A>.
```

* Preserves state using [`[PersistentState]` attribute](xref:Microsoft.AspNetCore.Components.PersistentStateAttribute) on component properties.

In the following example implementation, the following code files are placed in a `Shutdown` folder at the root of the app:

* `CircuitShutdownService.cs`: A singleton service that coordinates shutdown.
* `ShutdownCircuitHandler.cs`: A scoped service for each active circuit.
* `ShutdownCircuitOptions.cs`: Configuration options.

`Shutdown/ShutdownCircuitOptions.cs`:

```csharp
namespace PauseResumeOnShutdown.Shutdown;

public class ShutdownCircuitOptions
{
    public TimeSpan ShutdownTimeout { get; set; } = TimeSpan.FromSeconds(10);
}
```

`Shutdown/CircuitShutdownService.cs`:

```csharp

> using System.Collections.Concurrent;

Reviewer: ```suggestion
using System.Collections.Concurrent;
using Microsoft.AspNetCore.Components.Server.Circuits;
```

For `ConcurrentDictionary<string, Circuit>`.

using Microsoft.Extensions.Options;

namespace PauseResumeOnShutdown.Shutdown;

public class CircuitShutdownService
{

>     private readonly ConcurrentDictionary<string, ShutdownCircuitHandler> 

Reviewer: ```suggestion
    private readonly ConcurrentDictionary<string, Circuit> 
```

>         _handlers = new();

Reviewer: ```suggestion
        _circuits = new();
```

    private readonly ShutdownCircuitOptions _options;
    private bool _isShuttingDown;
    private TaskCompletionSource _shutdownTcs = new();

    public CircuitShutdownService(IHostApplicationLifetime appLifetime, 
        IOptions<ShutdownCircuitOptions> options)
    {
        _options = options.Value;
        appLifetime.ApplicationStopping.Register(OnApplicationStopping);
    }

    private void OnApplicationStopping()
    {
        _isShuttingDown = true;

>         if (_handlers.IsEmpty)

Reviewer: ```suggestion
        if (_circuits.IsEmpty)
```

        {
            return;
        }

        foreach (var handler in _handlers.Values)
        {

>             handler.Pause();

Reviewer: I don't think that we need the handler wrapper anymore. It's obscuring the view. Let's call directly:

```c#
var pauseTasks = _circuits.Values
            .Select(c => c.RequestCircuitPauseAsync().AsTask())
            .Append(_shutdownTcs.Task);
Task.WhenAll(pauseTasks).Wait(_options.ShutdownTimeout);
```

And maybe it would be good to mention somewhere that the fact that we're sending the `RequestCircuitPauseAsync` async doesn't mean that on returning the value the client already paused. It's only a request to pause that client can defer. That's why we have `_shutdownTcs` which is set when we don't have any circuits connected (all of them successfully shut down). In case a client requested a deferral longer than server wanted to allow (`>ShutdownTimeout`), that client won't have their state persisted and will experience a normal connection loss.

        }

        _shutdownTcs.Task.Wait(_options.ShutdownTimeout);
    }

>     public void Register(string circuitId, ShutdownCircuitHandler handler)

Reviewer: ```suggestion
    public void Register(string circuitId, Circuit circuit)
```

    {

>         _handlers.TryAdd(circuitId, handler);

Reviewer: ```suggestion
        _circuits.TryAdd(circuitId, circuit);
```

    }

    public void Unregister(string circuitId)
    {

>         _handlers.TryRemove(circuitId, out _);

Reviewer: ```suggestion
        _circuits.TryRemove(circuitId, out _);
```

>         if (_isShuttingDown && _handlers.IsEmpty)

Reviewer: ```suggestion
        if (_isShuttingDown && _circuits.IsEmpty)
```

        {
            _shutdownTcs.TrySetResult();
        }
    }
}
```

`Shutdown/ShutdownCircuitHandler.cs`:

```csharp
using Microsoft.AspNetCore.Components.Server.Circuits;

> using Microsoft.JSInterop;

Reviewer: ```suggestion
```

namespace PauseResumeOnShutdown.Shutdown;

public class ShutdownCircuitHandler(
    CircuitShutdownService shutdownService,

>     IJSRuntime jsRuntime) : CircuitHandler

Reviewer: ```suggestion
    ) : CircuitHandler
```
`IJSRuntime` was needed only when we didn't have `RequestCircuitPauseAsync`.

{
    public override Task OnConnectionUpAsync(Circuit circuit, 
        CancellationToken cancellationToken)
    {

>         shutdownService.Register(circuit.Id, this);

Reviewer: ```suggestion
        shutdownService.Register(circuit.Id, circuit);
```

        return Task.CompletedTask;
    }

    public override Task OnConnectionDownAsync(Circuit circuit, 
        CancellationToken cancellationToken)
    {
        shutdownService.Unregister(circuit.Id);

        return Task.CompletedTask;
    }

>     public void Pause()

Reviewer: `Pause` and `PauseCore` not necessary.

    {
        _ = PauseCore();
    }

    private async Task PauseCore()
    {
        try
        {

>             await jsRuntime.InvokeVoidAsync("remotePause");

Reviewer: this was the way to do `RequestCircuitPauseAsync` before we merged https://github.com/dotnet/aspnetcore/pull/66265.

        }
        catch
        {
        }
    }
}
```

`Program.cs`:

```csharp
using Microsoft.AspNetCore.Components.Server.Circuits;
using Microsoft.Extensions.DependencyInjection.Extensions;
using PauseResumeOnShutdown.Components;
using PauseResumeOnShutdown.Shutdown;

var builder = WebApplication.CreateBuilder(args);

// Increase host shutdown timeout to allow time for pause operations

> // Default value is 10 seconds

Reviewer: ```suggestion
// Must be greater than `ShutdownTimeout` in `ShutdownCircuitOptions` otherwise the
// host terminates connections before circuits finish pausing.
```

builder.Host.ConfigureHostOptions(options =>
    options.ShutdownTimeout = TimeSpan.FromSeconds(30));

> // Set circuit shutdown timeout to allow time for the host to restart

Reviewer: ```suggestion
```

The timeout has nothing to do with the host restarting.

> // Default value is 10 seconds per 'Shutdown/ShutdownCircuitOptions.cs'

Reviewer: ```suggestion
```

 redundant

builder.Services.Configure<ShutdownCircuitOptions>(options =>
    options.ShutdownTimeout = TimeSpan.FromSeconds(10));

builder.Services.AddRazorComponents()
    .AddInteractiveServerComponents();

// Register CircuitShutdownService as a singleton
builder.Services.AddSingleton<CircuitShutdownService>();

// Register ShutdownCircuitHandler as a scoped CircuitHandler
builder.Services.TryAddEnumerable(
    ServiceDescriptor.Scoped<CircuitHandler, ShutdownCircuitHandler>());

var app = builder.Build();

// ... rest of pipeline
```

> In `App.razor` after the [server-side Blazor script reference](xref:blazor/project-structure#location-of-the-blazor-script), `window.remotePause` is called from the server to trigger pause and returns immediately to avoid a "`Cannot send data`" error when Blazor attempts to send the response back.

Reviewer: This is about the workaround that could have been used before the PR that added pause-from-server API. Let's change it to a paragraph that actually describes how to delay the pause on client in cases when we know client is busy with work that cannot be interrupted (file transfer etc).

```suggestion
Optionally, to defer pause on the client until critical work completes (for example, an in-flight payment), configure the `onPauseRequested` callback in the [Blazor startup configuration](xref:blazor/fundamentals/startup). Place the following after the [server-side Blazor script reference](xref:blazor/project-structure#location-of-the-blazor-script):
```

```razor
<script> 

>   window.remotePause = function () {

Reviewer: All code inside of script should be raplaced with:

```suggestion
  Blazor.start({
    circuit: {
      onPauseRequested: async () => {
        // Perform any critical cleanup or wait for in-flight operations.
        // Return true to allow the pause, or false to reject it.
        return true;
      }
    }
  });
```

    Blazor.pauseCircuit();
  };
</script>
```

> The `remotePause` function must not be `async` and must not return a value. If it returns a `Promise`, Blazor attempts to send the result back after the connection closes, which results in an error.

Reviewer: ```suggestion
Without the `onPauseRequested` callback, the client pauses immediately when the server requests it.
```

        message = TempData?.Get("Message") as string ?? "No message";
    }

    private void Submit()
    {
        TempData!["Message"] = "Form submitted successfully!";

>         NavigationManager.NavigateTo("/tempdata-example", forceLoad: true);

Reviewer 2: ```suggestion
        NavigationManager.NavigateTo("/tempdata-example-1", forceLoad: true);
```
@guardrex, The TempDataExample1 component declares its route as @page "/tempdata-example-1" (line 302), but the Submit method navigates to "/tempdata-example" (without the -1 suffix).  Should this be `tempdata-example-1`?
